### PR TITLE
Optional Proof Support for View Access Key Endpoint

### PR DESCRIPTION
**Pull request by epociask**
_Author(s):_ @epociask

**View Access Key**
- Added support for a boolean `include_proof` query parameter for JSON API `view_access_key` method
- Extended RPC querying functionality and AccessKey primitives to support `Arc<Vec<8>>` proof type
- Generate _proof_ (ie merkle path) using Trie iterator that's configured to remember all visited Trie starting from root

**Testing**
Using default locally generated genesis state, queried `view_access_key` for `test.near` _account_id_ and `ed25519:2ZcmxtePPHt5fPioTCYxpv1VrEZH7i8kyGYNM1WGcYy2` for _public_key_ with the following request body:
```
curl --location --request POST 'http://0.0.0.0:8332' \
--header 'Content-Type: application/json' \
--data-raw '{
  "jsonrpc": "2.0",
  "id": "1",
  "method": "query",
  "params": {
    "request_type": "view_access_key",
    "finality": "final",
    "account_id": "test.near",
    "public_key": "ed25519:2ZcmxtePPHt5fPioTCYxpv1VrEZH7i8kyGYNM1WGcYy2",
    "include_proof": true
  }
}'
```

Which generated the following response:
```
{
"jsonrpc": "2.0",
"result:{
   "block_hash":"353Mo1n9vL9HRCJM4Yr9G8vCiVr1t1amDXV5LVWYstVi",
   "block_height": 21129,
   "nonce": 0,
   "permission": "FullAccess",
   "proof": [
 [3,1,0,0,0,16,192,245,53,162,217,209,135,151,181,55,146,182,95,167,70,52,53,13,190,245,212,144,217,10,107,226,58,43,134,133,214,13,184,3,0,0,0,0,0,0],
[1,5,0,244,175,27,70,82,220,235,64,177,19,169,163,2,221,117,205,105,65,190,117,231,173,7,175,227,205,49,150,33,113,234,98,34,184,116,138,208,55,48,160,155,16,196,197,71,78,60,231,222,190,112,86,172,1,148,57,155,47,70,2,233,170,91,15,132,3,0,0,0,0,0,0],
[1,192,0,30,155,206,148,95,192,254,181,166,67,60,154,212,48,236,29,211,108,34,171,225,193,123,171,25,16,183,157,192,212,238,11,231,252,112,147,28,143,28,3,167,30,87,105,95,101,124,187,11,69,157,12,239,62,246,15,160,49,36,23,151,9,91,189,174,1,0,0,0,0,0,0],
[0,43,0,0,0,52,101,115,116,46,110,101,97,114,2,0,23,54,185,88,81,164,59,206,241,95,158,132,163,106,251,101,64,164,68,13,183,12,235,253,59,112,216,15,112,6,7,149,9,0,0,0,42,225,193,156,12,189,55,142,70,201,39,169,243,97,25,35,236,7,204,26,227,87,80,42,9,83,109,69,82,117,207,33,195,0,0,0,0,0,0,0]
]
},
"id":"1"
}
```

### CODE REVIEW

File: core/primitives/src/trie_key.rs [hunk 1]
```diff
@@ -405,6 +405,19 @@ pub mod trie_key_parsers {
         res
     }
 
+    pub fn get_raw_prefix_for_access_key(
```
Comment (mm-near) [inline core/primitives/src/trie_key.rs:408]: You don't need this method. You can call the TrieKey.to_vec() instead. 

(the 'get_raw_prefix_for_access_keys' above was needed, as it was looking at a root for multiple keys - here you want to get the path to a concrete trie key)

File: core/primitives/src/views.rs [hunk 2]
```diff
@@ -226,6 +237,15 @@ pub struct StateItem {
     pub proof: Vec<()>,
 }
 
+#[derive(serde::Serialize, serde::Deserialize, Debug, PartialEq, Eq, Clone)]
+pub struct ViewAccessKeyResult {
+    pub access_key: AccessKey,
+    // TODO(mina86): Empty proof (i.e. sending proof when include_proof is not
+    // set in the request) was deprecated in 1.30.  Add
+    // `#[serde(skip(Vec::if_empty))` at 1.33 or something.
+    pub proof: Vec<Arc<[u8]>>,
```
Comment (mm-near) [inline core/primitives/src/views.rs:246]: Remove the TODO above (as this is not really true in this scenario).
Set this to Option<Vec<...>>

And add a comment that the proof is a serialized array of nodes that make up a merkle path from the key to the root of the shard.

File: chain/jsonrpc/jsonrpc-tests/tests/rpc_query.rs [hunk 3]
```diff
@@ -273,6 +273,7 @@ fn test_query_access_key() {
                     public_key: "ed25519:23vYngy8iL7q94jby3gszBnZ9JptpMf5Hgf7KVVa2yQ2"
                         .parse()
                         .unwrap(),
+                    include_proof: true,
                 },
             })
             .await
```
Comment (mm-near) [inline chain/jsonrpc/jsonrpc-tests/tests/rpc_query.rs:279]: could you also check that the proof is actually included in the respoinse?

File: core/primitives/src/views.rs [hunk 4]
```diff
@@ -251,6 +271,7 @@ pub struct QueryError {
 pub struct AccessKeyInfoView {
     pub public_key: PublicKey,
     pub access_key: AccessKeyView,
+    pub proof: Vec<Arc<[u8]>>,
```
Comment (mm-near) [inline core/primitives/src/views.rs:274]: this proof is never actually used and can be removed (the real proof is inside the AccessKeyView struct)

File: core/primitives/src/views.rs [hunk 5]
```diff
@@ -198,11 +198,22 @@ impl From<AccessKeyPermissionView> for AccessKeyPermission {
 pub struct AccessKeyView {
     pub nonce: Nonce,
     pub permission: AccessKeyPermissionView,
+    pub proof: Vec<Arc<[u8]>>,
```
Comment (mm-near) [inline core/primitives/src/views.rs:201]: Option<Vec<...>>

File: runtime/runtime/src/state_viewer/mod.rs [hunk 6]
```diff
@@ -77,19 +77,33 @@ impl TrieViewer {
...
         public_key: &PublicKey,
-    ) -> Result<AccessKey, errors::ViewAccessKeyError> {
-        get_access_key(state_update, account_id, public_key)?.ok_or_else(|| {
-            errors::ViewAccessKeyError::AccessKeyDoesNotExist { public_key: public_key.clone() }
-        })
+        include_proof: bool,
+    ) -> Result<ViewAccessKeyResult, errors::ViewAccessKeyError> {
+        let prefix = trie_key_parsers::get_raw_prefix_for_access_key(account_id, public_key);
+
+        let mut iter = state_update.trie().iter()?;
```
Comment (mm-near) [inline runtime/runtime/src/state_viewer/mod.rs:84]: we should create this iterator only if we are creating the proof.

File: runtime/runtime/src/state_viewer/mod.rs [hunk 7]
```diff
@@ -77,19 +77,33 @@ impl TrieViewer {
...
 
+    // TODO - Add boolean proof parameter support
     pub fn view_access_keys(
         &self,
         state_update: &TrieUpdate,
         account_id: &AccountId,
     ) -> Result<Vec<(PublicKey, AccessKey)>, errors::ViewAccessKeyError> {
         let prefix = trie_key_parsers::get_raw_prefix_for_access_keys(account_id);
         let raw_prefix: &[u8] = prefix.as_ref();
+
```
Comment (mm-near) [inline runtime/runtime/src/state_viewer/mod.rs:106]: remove

File: runtime/runtime/src/state_viewer/mod.rs [hunk 8]
```diff
@@ -77,19 +77,33 @@ impl TrieViewer {
...
+            get_access_key(state_update, account_id, public_key)?.ok_or_else(|| {
+                errors::ViewAccessKeyError::AccessKeyDoesNotExist { public_key: public_key.clone() }
+            })?;
+
+        let proof = iter.into_visited_nodes();
+
+        Ok(ViewAccessKeyResult { access_key, proof })
     }
 
+    // TODO - Add boolean proof parameter support
```
Comment (mm-near) [inline runtime/runtime/src/state_viewer/mod.rs:98]: if you're adding a TODO - please create a github issue and mention its number here.

File: runtime/runtime/src/state_viewer/mod.rs [hunk 9]
```diff
@@ -77,19 +77,33 @@ impl TrieViewer {
         state_update: &TrieUpdate,
         account_id: &AccountId,
         public_key: &PublicKey,
-    ) -> Result<AccessKey, errors::ViewAccessKeyError> {
-        get_access_key(state_update, account_id, public_key)?.ok_or_else(|| {
-            errors::ViewAccessKeyError::AccessKeyDoesNotExist { public_key: public_key.clone() }
-        })
+        include_proof: bool,
+    ) -> Result<ViewAccessKeyResult, errors::ViewAccessKeyError> {
+        let prefix = trie_key_parsers::get_raw_prefix_for_access_key(account_id, public_key);
```
Comment (mm-near) [inline runtime/runtime/src/state_viewer/mod.rs:82]: TrieKey::AccessKey(account_id, public_key).to_vec()

And this is not a prefix, but simply a key.

**Review by mm-near — CHANGES_REQUESTED** [review]
First - huge thank you for the PR. - especially as it is touching quite complex logic, so I'm really impressed.

The only thing I'm missing here is a test (and/or a library) to verify that the returned proof actually makes sense :-)

File: core/primitives/src/views.rs (hunk 4 above)
Comment (epociask) [reply core/primitives/src/views.rs:274]: Great callout

File: runtime/runtime/src/state_viewer/mod.rs (hunk 9 above)
Comment (epociask) [reply runtime/runtime/src/state_viewer/mod.rs]: [e2cded2](https://github.com/near/nearcore/pull/8644/commits/e2cded2a2c205f5cf907dd1e22d0ca2833a6a63f)

File: runtime/runtime/src/state_viewer/mod.rs (hunk 8 above)
Comment (epociask) [reply runtime/runtime/src/state_viewer/mod.rs]: [e2cded2](https://github.com/near/nearcore/pull/8644/commits/e2cded2a2c205f5cf907dd1e22d0ca2833a6a63f)

File: runtime/runtime/src/state_viewer/mod.rs (hunk 7 above)
Comment (epociask) [reply runtime/runtime/src/state_viewer/mod.rs:115]: [e2cded2](https://github.com/near/nearcore/pull/8644/commits/e2cded2a2c205f5cf907dd1e22d0ca2833a6a63f)

File: runtime/runtime/src/state_viewer/mod.rs (hunk 6 above)
Comment (epociask) [reply runtime/runtime/src/state_viewer/mod.rs]: [e2cded2](https://github.com/near/nearcore/pull/8644/commits/e2cded2a2c205f5cf907dd1e22d0ca2833a6a63f)

File: core/primitives/src/views.rs (hunk 5 above)
Comment (epociask) [reply core/primitives/src/views.rs]: [e2cded2](https://github.com/near/nearcore/pull/8644/commits/e2cded2a2c205f5cf907dd1e22d0ca2833a6a63f)

File: core/primitives/src/views.rs (hunk 2 above)
Comment (epociask) [reply core/primitives/src/views.rs]: [e2cded2](https://github.com/near/nearcore/pull/8644/commits/e2cded2a2c205f5cf907dd1e22d0ca2833a6a63f)

File: core/primitives/src/trie_key.rs (hunk 1 above)
Comment (epociask) [reply core/primitives/src/trie_key.rs]: [e2cded2](https://github.com/near/nearcore/pull/8644/commits/e2cded2a2c205f5cf907dd1e22d0ca2833a6a63f)

File: chain/jsonrpc/src/api/query.rs [hunk 10]
```diff
@@ -45,15 +45,26 @@ fn parse_path_data(path: String, data: String) -> Result<RpcQueryRequest, RpcPar
...
         .parse()
         .map_err(|err| RpcParseError(format!("{}", err)))?;
+
     let maybe_extra_arg = path_parts.next();
+    let maybe_require_proof: bool = match path_parts.next() {
+        Some(val) => val.contains(&String::from("true")),
+        None => false,
+    };
+
+    println!("Requires proof? {}", maybe_require_proof);
```
[the remark this answers is not in the document]
Comment (epociask) [reply chain/jsonrpc/src/api/query.rs]: [e2cded2](https://github.com/near/nearcore/pull/8644/commits/e2cded2a2c205f5cf907dd1e22d0ca2833a6a63f)

File: chain/jsonrpc/src/api/query.rs [hunk 11]
```diff
@@ -45,15 +45,26 @@ fn parse_path_data(path: String, data: String) -> Result<RpcQueryRequest, RpcPar
         parse_bs58_data(max_len, data)
     };
 
+    //command=>account_id=>access_key
```
[the remark this answers is not in the document]
Comment (epociask) [reply chain/jsonrpc/src/api/query.rs]: [e2cded2](https://github.com/near/nearcore/pull/8644/commits/e2cded2a2c205f5cf907dd1e22d0ca2833a6a63f)

**Comment by epociask** [comment]
@mm-near Just a heads up that this PR got all screwed up during a rebase, forked off branch and opened other PR here: https://github.com/near/nearcore/pull/8660

**Pull request closed without merging**